File: packages/agent-runtime/src/runtime.ts
/**
 * @ainative/agent-runtime — AgentRuntime
 * Built by AINative Dev Team
 * Refs #246
 */

import type {
  RuntimeConfig,
  AgentTask,
  TurnResult,
  RunResult,
  RunStatus,
  StepContext,
  LLMMessage,
  ToolCall,
  Tool,
} from './types';

// ─── Event Types ─────────────────────────────────────────────────────────────

export type RuntimeEvent =
  | 'turn_start'
  | 'turn_end'
  | 'tool_call'
  | 'tool_result'
  | 'complete'
  | 'error';

export type EventCallback = (data: unknown) => void;

// ─── AgentRuntime ────────────────────────────────────────────────────────────

/**
 * Embeddable agent runtime that executes multi-turn agent loops.
 * Works in Node.js, Electron, Tauri, or any environment with async/await.
 */
export class AgentRuntime {
  readonly maxTurns: number;

  private readonly config: RuntimeConfig;
  private readonly listeners: Map<RuntimeEvent, EventCallback[]> = new Map();
  private readonly globalTools: Tool[];

  constructor(config: RuntimeConfig) {
    this.config = config;
    this.maxTurns = config.maxTurns ?? 10;
    this.globalTools = config.tools ?? [];
  }

  // ─── Event Emitter ────────────────────────────────────────────────────────

  on(event: RuntimeEvent, callback: EventCallback): void {
    const existing = this.listeners.get(event) ?? [];
    this.listeners.set(event, [...existing, callback]);
  }

  private emit(event: RuntimeEvent, data: unknown): void {
    const callbacks = this.listeners.get(event) ?? [];
    for (const cb of callbacks) {
      cb(data);
    }
  }

  // ─── step() ───────────────────────────────────────────────────────────────

  /**
   * Execute a single agent turn: think → select tool → execute → record.
   */
  async step(context: StepContext): Promise<TurnResult> {
    const { messages, tools, options } = context;

    const llmToolDefs = tools.map((t) => ({
      name: t.name,
      description: t.description,
      parameters: t.parameters,
    }));

    const llmResponse = await this.config.llmProvider.chatWithTools(
      messages,
      llmToolDefs,
      options,
    );

    const toolCalls: ToolCall[] = [];

    for (const tc of llmResponse.toolCalls) {
      this.emit('tool_call', tc);

      const tool = tools.find((t) => t.name === tc.name);
      const executed: ToolCall = { ...tc };

      if (tool) {
        try {
          executed.result = await tool.execute(tc.args);
        } catch (err) {
          executed.error = err instanceof Error ? err.message : String(err);
        }
      } else {
        executed.error = `Tool "${tc.name}" not found`;
      }

      this.emit('tool_result', executed);
      toolCalls.push(executed);
    }

    return {
      turnNumber: 0, // caller fills in the turn number
      thought: llmResponse.content,
      toolCalls,
      messages,
    };
  }

  // ─── run() ────────────────────────────────────────────────────────────────

  /**
   * Execute the full agent loop for a task, up to maxTurns.
   */
  async run(task: AgentTask): Promise<RunResult> {
    const allTools = [...this.globalTools, ...task.tools];
    const turns: TurnResult[] = [];

    const systemMessage: LLMMessage = {
      role: 'system',
      content: task.systemPrompt ?? 'You are a helpful AI assistant. Complete the given task.',
    };
    const userMessage: LLMMessage = { role: 'user', content: task.description };

    const messages: LLMMessage[] = [systemMessage, userMessage];

    let status: RunStatus = 'complete';
    let finalAnswer: string | undefined;
    let errorMessage: string | undefined;

    try {
      for (let turn = 0; turn < this.maxTurns; turn++) {
        this.emit('turn_start', { turn, taskId: task.id });

        const turnResult = await this.step({ messages, tools: allTools });
        turnResult.turnNumber = turn + 1;

        turns.push(turnResult);
        this.emit('turn_end', turnResult);

        // Append assistant thought to conversation
        if (turnResult.thought) {
          messages.push({ role: 'assistant', content: turnResult.thought });
        }

        // Append tool results to conversation
        for (const tc of turnResult.toolCalls) {
          messages.push({
            role: 'tool',
            content: tc.error
              ? `Error: ${tc.error}`
              : String(tc.result ?? ''),
            toolCallId: tc.id,
          });
        }

        // If LLM returned no tool calls and has content, the agent is done
        if (turnResult.toolCalls.length === 0 && turnResult.thought) {
          finalAnswer = turnResult.thought;
          status = 'complete';
          break;
        }

        // If this is the last allowed turn and we still have tool calls, mark limit
        if (turn === this.maxTurns - 1) {
          status = 'max_turns_reached';
        }
      }
    } catch (err) {
      status = 'error';
      errorMessage = err instanceof Error ? err.message : String(err);
      this.emit('error', { taskId: task.id, error: errorMessage });
    }

    const result: RunResult = {
      taskId: task.id,
      status,
      turns,
      finalAnswer,
      error: errorMessage,
    };

    // Persist to storage
    try {
      await this.config.storage.storeRecord('agent_runs', {
        taskId: task.id,
        status,
        turnsCount: turns.length,
        finalAnswer,
        error: errorMessage,
        completedAt: new Date().toISOString(),
      });
    } catch {
      // Storage errors do not fail the run
    }

    if (status !== 'error') {
      this.emit('complete', result);
    }

    return result;
  }
}

File: packages/agent-runtime-py/ainative_agent_runtime/runtime.py
"""
ainative-agent-runtime — AgentRuntime
Built by AINative Dev Team
Refs #246

Embeddable agent runtime for Python. Mirrors the TypeScript AgentRuntime.
"""

from __future__ import annotations

import asyncio
from collections import defaultdict
from typing import Any, Callable, Dict, List, Optional


# Event type
RuntimeEvent = str
EventCallback = Callable[[Any], None]


class AgentRuntime:
    """
    Embeddable agent runtime that executes multi-turn agent loops.

    Args:
        storage: A StorageAdapter instance (local or cloud).
        llm_provider: An LLMProvider instance.
        tools: Optional list of tool definitions (dicts with name, description, execute).
        max_turns: Maximum number of turns before halting (default: 10).
    """

    def __init__(
        self,
        storage: Any,
        llm_provider: Any,
        tools: Optional[List[Dict[str, Any]]] = None,
        max_turns: int = 10,
    ) -> None:
        self._storage = storage
        self._llm = llm_provider
        self._global_tools: List[Dict[str, Any]] = tools or []
        self.max_turns = max_turns
        self._listeners: Dict[str, List[EventCallback]] = defaultdict(list)

    # ─── Event Emitter ────────────────────────────────────────────────────────

    def on(self, event: RuntimeEvent, callback: EventCallback) -> None:
        """Register a listener for the given event."""
        self._listeners[event].append(callback)

    def _emit(self, event: RuntimeEvent, data: Any) -> None:
        for cb in self._listeners.get(event, []):
            cb(data)

    # ─── step() ───────────────────────────────────────────────────────────────

    async def step(self, context: Dict[str, Any]) -> Dict[str, Any]:
        """
        Execute a single agent turn: think → select tool → execute → record.

        Args:
            context: dict with keys 'messages', 'tools', optionally 'options'.

        Returns:
            A dict representing a TurnResult.
        """
        messages = context.get("messages", [])
        tools: List[Dict[str, Any]] = context.get("tools", [])
        options = context.get("options", None)

        # Build LLM tool definitions (strip execute callable)
        llm_tool_defs = [
            {"name": t["name"], "description": t["description"], "parameters": t.get("parameters")}
            for t in tools
        ]

        llm_response = await self._llm.chat_with_tools(messages, llm_tool_defs, options)

        tool_calls_result = []
        for tc in llm_response.get("tool_calls", []):
            self._emit("tool_call", tc)

            tool = next((t for t in tools if t["name"] == tc["name"]), None)
            executed = dict(tc)

            if tool:
                try:
                    executed["result"] = await tool["execute"](tc.get("args", {}))
                except Exception as exc:
                    executed["error"] = str(exc)
            else:
                executed["error"] = f'Tool "{tc["name"]}" not found'

            self._emit("tool_result", executed)
            tool_calls_result.append(executed)

        return {
            "turn_number": 0,  # caller updates this
            "thought": llm_response.get("content", ""),
            "tool_calls": tool_calls_result,
            "messages": messages,
        }

    # ─── run() ────────────────────────────────────────────────────────────────

    async def run(self, task: Dict[str, Any]) -> Dict[str, Any]:
        """
        Execute the full agent loop for a task, up to max_turns.

        Args:
            task: dict with id, description, tools, metadata.

        Returns:
            A dict representing a RunResult.
        """
        all_tools = self._global_tools + task.get("tools", [])
        turns = []
        task_id = task.get("id", "unknown")

        messages = [
            {"role": "system", "content": task.get("system_prompt", "You are a helpful AI assistant.")},
            {"role": "user", "content": task.get("description", "")},
        ]

        status = "complete"
        final_answer: Optional[str] = None
        error_message: Optional[str] = None

        try:
            for turn_num in range(self.max_turns):
                self._emit("turn_start", {"turn": turn_num, "task_id": task_id})

                turn_result = await self.step({"messages": messages, "tools": all_tools})
                turn_result["turn_number"] = turn_num + 1
                turns.append(turn_result)
                self._emit("turn_end", turn_result)

                thought = turn_result.get("thought", "")
                tool_calls = turn_result.get("tool_calls", [])

                if thought:
                    messages.append({"role": "assistant", "content": thought})

                for tc in tool_calls:
                    messages.append({
                        "role": "tool",
                        "content": f"Error: {tc['error']}" if tc.get("error") else str(tc.get("result", "")),
                        "tool_call_id": tc.get("id"),
                    })

                # LLM produced text and no tool calls → done
                if not tool_calls and thought:
                    final_answer = thought
                    status = "complete"
                    break

                if turn_num == self.max_turns - 1:
                    status = "max_turns_reached"

        except Exception as exc:
            status = "error"
            error_message = str(exc)
            self._emit("error", {"task_id": task_id, "error": error_message})

        result = {
            "task_id": task_id,
            "status": status,
            "turns": turns,
            "final_answer": final_answer,
            "error": error_message,
        }

        # Persist
        try:
            await self._storage.store_record("agent_runs", {
                "task_id": task_id,
                "status": status,
                "turns_count": len(turns),
                "final_answer": final_answer,
                "error": error_message,
            })
        except Exception:
            pass  # Storage errors do not fail the run

        if status != "error":
            self._emit("complete", result)

        return result
